classes/village/VillageManager.php

```diff
@@ -1708,6 +1726,72 @@ public static function enactProposal(System $system, User $player, int $proposal
                 self::createProposalNotification($system, $proposal['village_id'], NotificationManager::NOTIFICATION_PROPOSAL_PASSED, $proposal['name']);
                 $message = "Accepted trade offer from " . VillageManager::VILLAGE_NAMES[$proposal['target_village_id']] . "!";
                 break;
+            case self::PROPOSAL_TYPE_BEGIN_CONSTRUCTION:
+                try {
+                    $message = VillageUpgradeManager::beginConstruction($system, $player->village, $proposal['building_key']);
+                    // update proposal
+                    $system->db->query("UPDATE `proposals` SET `end_time` = {$time}, `result` = 'passed' WHERE `proposal_id` = {$proposal['proposal_id']}");
+                    // create notification
+                    self::createProposalNotification($system, $proposal['village_id'], NotificationManager::NOTIFICATION_PROPOSAL_PASSED, $proposal['name']);
+                } catch (RuntimeException $e) {
+                    return $e->getMessage();
+                }
```

Comment:
This logic is duplicated for every proposal in here and this function is getting super long - Could you move the try/catch above the switch statement, and also the lines for updating the proposal in DB and creating notification? 

Could also make the proposal implementation functions like `VillageUpgradeManager::beginConstruction` return an instance of `ActionResult` and then you can handle success/failure in a standard way

Reply:
I'm planning on taking a pass at proposal logic for this reason. There's a lot of code reuse and VillageManager is getting bloated, so I want to make a ProposalManager class to handle all of this logic and cut down on repeated actions.

For ActionResult vs Try/Catch with a String return, which do you think is better? ActionResult would have a string message and bool success. Whichever we choose I would like to apply to the other functions here.